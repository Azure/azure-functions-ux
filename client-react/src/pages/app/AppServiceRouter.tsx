import React, { useEffect, useState, lazy, useContext } from 'react';
import { RouteComponentProps, Router } from '@reach/router';
import { StartupInfoContext } from '../../StartupInfoContext';
import { iconStyles } from '../../theme/iconStyles';
import { ThemeContext } from '../../ThemeContext';
export interface AppSeriviceRouterProps {
  subscriptionId?: string;
  resourcegroup?: string;
  siteName?: string;
  slotName?: string;
  functionName?: string;
}

const AppSettingsLoadable: any = lazy(() => import(/* webpackChunkName:"appsettings" */ './app-settings/AppSettings'));
const LogStreamLoadable: any = lazy(() => import(/* webpackChunkName:"logstream" */ './log-stream/LogStreamDataLoader'));
const ChangeAppPlanLoadable: any = lazy(() => import(/* webpackChunkName:"changeappplan" */ './change-app-plan/ChangeAppPlanDataLoader'));
const FunctionIntegrateLoadable: any = lazy(() =>
  import(/* webpackChunkName:"functionintegrate" */ './functions/integrate/FunctionIntegrateDataLoader')
);
const FunctionBindingEditorLoadable: any = lazy(() =>
  import(/* webpackChunkName:"functionbindingeditor" */ './functions/integrate/binding-editor/BindingEditorDataLoader')
);
const FunctionCreateLoadable: any = lazy(() =>
  import(/* webpackChunkName:"functioncreate" */ './functions/create/FunctionCreateDataLoader')
);

const AppServiceRouter: React.FC<RouteComponentProps<AppSeriviceRouterProps>> = props => {
  const [resourceId, setResourceId] = useState('');
  const theme = useContext(ThemeContext);

  useEffect(() => {
    const { subscriptionId, resourcegroup, siteName, slotName, functionName } = props;
    let id = `/subscriptions/${subscriptionId}/resourcegroups/${resourcegroup}/providers/Microsoft.Web/sites/${siteName}`;
    if (slotName) {
      id = `${id}/slots/${slotName}`;
    }

    if (functionName) {
      id = `${id}/functions/${functionName}`;
    }

    setResourceId(id);
  }, []);

  return (
    <main className={iconStyles(theme)}>
      <StartupInfoContext.Consumer>
        {value => {
          return (
            value.token && (
              <Router>
                <AppSettingsLoadable resourceId={resourceId} path="/settings" />
                <LogStreamLoadable resourceId={resourceId} path="/log-stream" />
                <ChangeAppPlanLoadable resourceId={resourceId} path="/changeappplan" />
                <FunctionIntegrateLoadable resourceId={resourceId} path="/integrate" />
                <FunctionBindingEditorLoadable resourceId={resourceId} path="/bindingeditor" />
                <FunctionCreateLoadable resourceId={resourceId} path="/functioncreate" />
              </Router>
            )
          );
        }}
      </StartupInfoContext.Consumer>
    </main>
  );
};
export default AppServiceRouter;
