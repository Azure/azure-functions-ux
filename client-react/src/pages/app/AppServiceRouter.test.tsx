import { shallow } from 'enzyme';
import * as React from 'react';
import { AppServiceRouter } from './AppServiceRouter';

describe('App Service Router', () => {
  const mockUpdateResourceIdCall = jest.fn();
  const subId = 'testsub';
  const rg = 'testrg';
  const site = 'sitename';
  const slot = 'slotname';
  describe('sites', () => {
    let container;
    beforeEach(() => {
      container = shallow(
        <AppServiceRouter updateResourceId={mockUpdateResourceIdCall} subscriptionId={subId} resourcegroup={rg} siteName={site} />
      );
    });
    afterEach(() => {
      mockUpdateResourceIdCall.mockReset();
    });
    it('renders without crashing', () => {
      expect(container.length).toEqual(1);
    });

    it('should render correctly', () => {
      expect(container).toMatchSnapshot('site-name-only');
    });

    it('calls update resourceId', () => {
      expect(mockUpdateResourceIdCall).toHaveBeenCalledWith(
        `/subscriptions/${subId}/resourcegroups/${rg}/providers/Microsoft.Web/sites/${site}`
      );
    });
  });

  describe('slots', () => {
    let container;
    beforeEach(() => {
      container = shallow(
        <AppServiceRouter
          updateResourceId={mockUpdateResourceIdCall}
          subscriptionId={subId}
          resourcegroup={rg}
          siteName={site}
          slotName={slot}
        />
      );
    });
    afterEach(() => {
      mockUpdateResourceIdCall.mockReset();
    });
    it('renders without crashing', () => {
      expect(container.length).toEqual(1);
    });

    it('should render correctly', () => {
      expect(container).toMatchSnapshot('site-and-slot-only');
    });

    it('calls update resourceId', () => {
      expect(mockUpdateResourceIdCall).toHaveBeenCalledWith(
        `/subscriptions/${subId}/resourcegroups/${rg}/providers/Microsoft.Web/sites/${site}/slots/${slot}`
      );
    });
  });
});
