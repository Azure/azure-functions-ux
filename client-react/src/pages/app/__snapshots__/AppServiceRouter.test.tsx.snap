// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Service Router sites should render correctly: site-name-only 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppServiceRouter
    resourcegroup="testrg"
    siteName="sitename"
    subscriptionId="testsub"
    updateResourceId={
      [MockFunction] {
        "calls": Array [
          Array [
            "/subscriptions/testsub/resourcegroups/testrg/providers/Microsoft.Web/sites/sitename",
          ],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <LoadableComponent
        path="/settings"
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "path": "/settings",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <LoadableComponent
          path="/settings"
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "path": "/settings",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`App Service Router slots should render correctly: site-and-slot-only 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppServiceRouter
    resourcegroup="testrg"
    siteName="sitename"
    slotName="slotname"
    subscriptionId="testsub"
    updateResourceId={
      [MockFunction] {
        "calls": Array [
          Array [
            "/subscriptions/testsub/resourcegroups/testrg/providers/Microsoft.Web/sites/sitename/slots/slotname",
          ],
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <LoadableComponent
        path="/settings"
      />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "path": "/settings",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <LoadableComponent
          path="/settings"
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "path": "/settings",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
